main.go

```diff
@@ -178,90 +193,207 @@ func main() {
+	flag.DurationVar(&opts.Period, "period", 5*time.Second, "The time to wait between remote-write requests.")
+	flag.DurationVar(&opts.Duration, "duration", 5*time.Minute, "The duration of the up command to run until it stops.")
+	flag.Float64Var(&opts.SuccessThreshold, "threshold", 0.9, "The percentage of successful requests needed to succeed overall. 0 - 1.")
+	flag.DurationVar(&opts.Latency, "latency", 15*time.Second, "The maximum allowable latency between writing and reading.")
+	flag.DurationVar(&opts.InitialQueryDelay, "initial-query-delay", 5*time.Second, "The time to wait before executing the first query.")
+	flag.Parse()
+
+	switch rawLogLevel {
+	case "error":
+		opts.LogLevel = level.AllowError()
+	case "warn":
+		opts.LogLevel = level.AllowWarn()
+	case "info":
+		opts.LogLevel = level.AllowInfo()
+	case "debug":
+		opts.LogLevel = level.AllowDebug()
+	default:
+		panic("unexpected log level")
+	}
+
+	writeEndpoint, err := url.ParseRequestURI(rawWriteEndpoint)
+	if err != nil {
+		return opts, fmt.Errorf("--endpoint-write is invalid: %w", err)
+	}
+
+	opts.WriteEndpoint = writeEndpoint
+
+	var readEndpoint *url.URL
+	if rawReadEndpoint != "" {
+		readEndpoint, err = url.ParseRequestURI(rawReadEndpoint)
+		if err != nil {
+			return opts, fmt.Errorf("--endpoint-read is invalid: %w", err)
+		}
+	}
+
+	opts.ReadEndpoint = readEndpoint
+
+	if opts.Duration <= opts.Period {
+		return opts, errors.New("--duration cannot be less than period")
+	}
+
+	if opts.Latency <= opts.Period {
+		return opts, errors.New("--latency cannot be less than period")
 	}
+
+	opts.Labels = append(opts.Labels, prompb.Label{
+		Name:  "__name__",
+		Value: opts.Name,
+	})
+
+	return opts, err
 }
 
-func read(ctx context.Context, endpoint *url.URL, labels []prompb.Label, latency time.Duration) error {
+func read(ctx context.Context, endpoint *url.URL, labels []prompb.Label, ago, latency time.Duration, m metrics) error {
 	client, err := promapi.NewClient(promapi.Config{Address: endpoint.String()})
 	if err != nil {
 		return err
 	}
-	a := promv1.NewAPI(client)
 
-	var labelSelectors []string
-	for _, label := range labels {
-		labelSelectors = append(labelSelectors, fmt.Sprintf(`%s="%s"`, label.Name, label.Value))
+	labelSelectors := make([]string, len(labels))
+	for i, label := range labels {
+		labelSelectors[i] = fmt.Sprintf(`%s="%s"`, label.Name, label.Value)
 	}
+
 	query := fmt.Sprintf("{%s}", strings.Join(labelSelectors, ","))
 
-	value, _, err := a.Query(ctx, query, time.Now().Add(-5*time.Second))
+	q := endpoint.Query()
+	q.Set("query", query)
+
+	ts := time.Now().Add(ago)
+	if !ts.IsZero() {
+		q.Set("time", formatTime(ts))
+	}
+
+	_, body, _, err := promapi.DoGetFallback(client, ctx, endpoint, q) //nolint:bodyclose
 	if err != nil {
-		return err
+		return errors.Wrap(err, "query request failed")
 	}
 
-	vec := value.(model.Vector)
+	var result queryResult
+	if err := json.Unmarshal(body, &result); err != nil {
```

Comment:
why are we doing this now vs using the prometheus client, which returns typed results?

Reply:
API Client library embeds URL paths so we can't specify arbitrarily paths to query against.
https://github.com/prometheus/client_golang/blob/b68ba26e470165fff70efc9cf97254a1544fa5b3/api/prometheus/v1/api.go#L661

I had to do this in order to specify arbitrary query urls.

Reply:
This was the reason why I changed `read-endpoint` to `read-address` and added a note that says you can only specify base prometheus url with this parameter temporarily. This is the fix for that.